src/tracing.rs

```diff
@@ -64,13 +40,551 @@ pub fn setup(user_filters: &str) -> anyhow::Result<()> {
+
+        pub fn prepare_update(&self) -> Option<Update> {
+            if self.has_changes() {
+                Some(Update::from_diff(self))
+            } else {
+                None
+            }
+        }
+    }
+
+    #[rustfmt::skip] // `rustfmt` formats this into unreadable single line
+    pub type Logger<S> = filter::Filtered<
+        Option<fmt::Layer<S>>,
+        filter::EnvFilter,
+        S,
+    >;
+
+    pub fn new<S>(config: &Config) -> Logger<S>
+    where
+        S: tracing::Subscriber + for<'span> registry::LookupSpan<'span>,
+    {
+        let layer = fmt::Layer::default()
+            .with_ansi(config.color.to_bool())
+            .with_span_events(config.span_events.clone().into());
+
+        let filter = filter(config.log_level.as_deref().unwrap_or(""));
+
+        Some(layer).with_filter(filter)
+    }
+
+    #[derive(Debug, Default)]
+    pub struct Update {
+        filter: Option<filter::EnvFilter>,
+        span_events: Option<fmt::format::FmtSpan>,
+        ansi: Option<bool>,
+    }
+
+    impl Update {
+        pub fn from_diff(diff: &ConfigDiff) -> Self {
+            let mut update = Self::default();
+            update.prepare(diff);
+            update
+        }
+
+        fn prepare(&mut self, diff: &ConfigDiff) {
+            if let Some(log_level) = &diff.log_level {
+                self.filter = Some(filter(log_level.as_deref().unwrap_or("")));
+            }
+
+            if let Some(span_events) = diff.span_events.clone() {
+                self.span_events = Some(span_events.into())
+            }
+
+            if let Some(color) = diff.color {
+                self.ansi = Some(color.to_bool());
+            }
+        }
+
+        // `apply` should *never* log anything or produce any `tracing` events!
+        pub fn apply<S>(self, logger: &mut Logger<S>) {
+            if let Some(filter) = self.filter {
+                *logger.filter_mut() = filter;
+            }
+
+            if let Some(span_events) = self.span_events {
+                let mut layer = logger.inner_mut().take().expect("valid logger state");
+                layer = layer.with_span_events(span_events);
+                *logger.inner_mut() = Some(layer);
+            }
+
+            if let Some(ansi) = self.ansi {
+                logger
+                    .inner_mut()
+                    .as_mut()
+                    .expect("valid logger state")
+                    .set_ansi(ansi);
+            }
+        }
+    }
+
+    fn filter(user_filters: &str) -> filter::EnvFilter {
+        const DEFAULT_LOG_LEVEL: log::LevelFilter = log::LevelFilter::Info;
+
+        const DEFAULT_FILTERS: &[(&str, log::LevelFilter)] = &[
+            ("hyper", log::LevelFilter::Info),
+            ("h2", log::LevelFilter::Error),
+            ("tower", log::LevelFilter::Warn),
+            ("rustls", log::LevelFilter::Info),
+            ("wal", log::LevelFilter::Warn),
+            ("raft", log::LevelFilter::Warn),
+        ];
```

Comment:
Nitpick, but shall we move these to the top of the file so they are a bit more visible?